presto-hive/src/main/java/com/facebook/presto/hive/HiveColumnHandle.java

```diff
@@ -52,6 +52,11 @@ public class HiveColumnHandle
     public static final HiveType BUCKET_HIVE_TYPE = HIVE_INT;
     public static final TypeSignature BUCKET_TYPE_SIGNATURE = BUCKET_HIVE_TYPE.getTypeSignature();
 
+    public static final int ROW_ID_COLUMN_INDEX = -16;
+    public static final String ROW_ID_COLUMN_NAME = "$row_id";
```

Comment:
Let's discuss the interface tomorrow. 

The reader itself can provide the row number, the file name (aka row_group) can be either provided by the reader (we do it already for the selective reader) or by the page source.
The row id is multiple things including the row number and the file name combined together. 

Ideally we would like to avoid making the file reader produce a single column with the row id because it will require the reader to 1. concatenate a lot of info and thus produce a lot of small varchar blobs, 2. and thinking about the wider context we don't want the reader to know about the row id concept. 

In my opinion the ideal situation would be if the file reader could return just the row number and may be a constant row group id (file name), and the rest of the stuff should happen on a higher level.

Reply:
Can't the file name just use the `$path` hidden column in Hive? https://prestodb.io/docs/current/connector/hive.html#extra-hidden-columns

Reply:
$path is the full path, in case of Metalake we need just the file name.

Reply:
> Ideally we would like to avoid making the file reader produce a single column with the row id because it will require the reader to 1. concatenate a lot of info and thus produce a lot of small varchar blobs, 2. and thinking about the wider context we don't want the reader to know about the row id concept.
> 

Actually, I would like ROW_ID to be of type ROWID type and maybe even make it not writeable to hive (and don't allow cast to any other type including varbinary). And for simplicity I want the "$row_id" to be returned by metastore, not built in presto. This will make it fool-proof between engines. 

> In my opinion the ideal situation would be if the file reader could return just the row number and may be a constant row group id (file name), and the rest of the stuff should happen on a higher level.

I'm not sure this is a good idea. We want the interface to be self-contained.

Reply:
@sdruzkin I thought you (or Brian?) had a doc that said we will make it a varbinary or somthing? then we can just make that the ROW_ID right?

Reply:
> $path is the full path, in case of Metalake we need just the file name.

If this is a Metalake-only concept, can you just add it to the PrismConnector?  I am not sure if there is an established convention on how to define a row in OSS Hive.

Reply:
> @sdruzkin I thought you (or Brian?) had a doc that said we will make it a varbinary or somthing? then we can just make that the ROW_ID right?

Yes, however the global ROWID proposed in the Brian's doc has a lot of things going on which are outside of the simple reader scope - it has the table and partition ids, data version, etc. These values are just outside of the reader's context. But we can built it on top of the values like the row number and row group id (file name), may be wrapping them into the ROWID type, that the reader can return.


> Actually, I would like ROW_ID to be of type ROWID type and maybe even make it not writeable to hive (and don't allow cast to any other type including varbinary). And for simplicity I want the "$row_id" to be returned by metastore, not built in presto. This will make it fool-proof between engines.

Is this going to be a new type? I cannot find an already existing type with this name in the OSS.

Reply:
Actually the doc got some major updates since I read it last time. But my statement about the scope of reader's responsibility still holds. We can model the reader's interface in a way where Presto/Spark provide some function as c-tor argument converting the row number to the global row id.

Reply:
So the Hive reader provides the row number and row group ID parts, Metastore provides the table ID, and then something (what?) concatenates them. Or do we pass the table ID part into the HIve reader which then concatenates it with the row number and row group ID?

FYI, either way, I'm still trying to figure out where in the code the Hive reader/connector is going to stuff in the row number. If anyone happens to know, a hint would be appreciated. :-)